File: hazelcastcsharpclientproblem/clusterandjavaproducer/hazelcastproblemdotnet/common/src/main/java/com/mycompany/common/TopicData.java
package com.mycompany.common;

import java.io.IOException;

import com.hazelcast.nio.serialization.Portable;
import com.hazelcast.nio.serialization.PortableReader;
import com.hazelcast.nio.serialization.PortableWriter;

public class TopicData implements Portable {
    
    private String           data;

    public TopicData() {

    }

    public TopicData(String data) {
        this.data = data;
    }

    @Override
    public int getFactoryId() {
        return 1;
    }

    @Override
    public int getClassId() {
        return 1;
    }
    
    @Override
    public String toString() {
        return "TopicData [data=" + data + "]";
    }

    @Override
    public void writePortable(PortableWriter writer) throws IOException {
        boolean hasData = (data != null);
        if (hasData) {
            writer.writeUTF("data", data);
        }
        
        writer.writeBoolean("_has__data", hasData);
        
    }

    @Override
    public void readPortable(PortableReader reader) throws IOException {
        if (reader.readBoolean("_has__data")) {
            data = reader.readUTF("data");
        }
        
    }

}

File: hazelcastcsharpclientproblem/dotnetclienttopiclistener/HazelcastDotNetProblem/TopicData.cs
using Hazelcast.IO.Serialization;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Runtime.Serialization;
using System.Text;
using System.Threading.Tasks;

namespace HazelcastDotNetProblem
{
    public class TopicData : IPortable, ISerializable
    {
        public String Data { get; set; }

        public int GetClassId()
        {
            return 1;
        }

        public int GetFactoryId()
        {
            return 1;
        }

        public void ReadPortable(IPortableReader reader)
        {
            if (reader.ReadBoolean("_has__data"))
            {
                Data = reader.ReadUTF("data");
            }



        }

        public void WritePortable(IPortableWriter writer)
        {
            bool hasData = (Data != null);
            if (hasData)
            {
                writer.WriteUTF("data", Data);
            }
            writer.WriteBoolean("_has__data", hasData);


        }


        public override String ToString()
        {
            return "TopicData [Data=" + Data + "]";
        }

        public void GetObjectData(SerializationInfo info, StreamingContext context)
        {
            // throw new NotImplementedException();
        }
    }
}
